### ISSUE TITLE
Potential racing issue on agent/activity creation, vs multiple statements issues at once

### ISSUE THREAD

**Issue by BugVito** [adlnet/ADL_LRS#368]
We are getting into an issue where an agent or activity is attempted to be created more than once, by sending a bunch of statements asynchronously.

From our tests, we assume that a check is made to see if the agent/activity should be created, but by the time the agent/activity is actually created, a few requests to create it has gone through to a queue.

This causes statements to get lost as the system refuses to create a duplicate agent/activity:

```
Exception while processing: /xAPI/statements
2015-11-10 15:35:25,771 [ERROR] lrs.views: duplicate key value violates unique constraint "lrs_activity_activity_id_key"
DETAIL:  Key (activity_id)=(http://blahblah/someactivity) already exists.
Traceback (most recent call last):
  File "/home/adllrs/ADL_LRS/lrs/views.py", line 619, in handle_request
    return processors[path][req_dict['method']](req_dict)
  File "/home/adllrs/ADL_LRS/lrs/util/req_process.py", line 124, in statements_post
    stmt_responses = process_body(body, auth, req_dict['headers']['X-Experience-API-Version'], req_dict.get('payload_sha2s', None))
  File "/home/adllrs/ADL_LRS/lrs/util/req_process.py", line 57, in process_body
    return [process_statement(st, auth, version, payload_sha2s) for st in stmts]
  File "/home/adllrs/ADL_LRS/lrs/util/req_process.py", line 50, in process_statement
    st = StatementManager(stmt, auth, payload_sha2s).model_object
  File "/home/adllrs/ADL_LRS/lrs/managers/StatementManager.py", line 15, in __init__
    self.populate(auth_info, stmt_data, payload_sha2s)
  File "/home/adllrs/ADL_LRS/lrs/managers/StatementManager.py", line 166, in populate
    self.build_statement_object(auth_info, stmt_data)
  File "/home/adllrs/ADL_LRS/lrs/managers/StatementManager.py", line 152, in build_statement_object
    define=auth_info['define']).Activity
  File "/home/adllrs/ADL_LRS/lrs/managers/ActivityManager.py", line 7, in __init__
    self.populate(data)
  File "/home/adllrs/ADL_LRS/lrs/managers/ActivityManager.py", line 63, in populate
    self.Activity = Activity.objects.create(activity_id=activity_id, authority=self.auth)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/manager.py", line 137, in create
    return self.get_query_set().create(**kwargs)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 377, in create
    obj.save(force_insert=True, using=self.db)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 463, in save
    self.save_base(using=using, force_insert=force_insert, force_update=force_update)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 551, in save_base
    result = manager._insert([self], fields=fields, return_id=update_pk, using=using, raw=raw)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/manager.py", line 203, in _insert
    return insert_query(self.model, objs, fields, **kwargs)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 1576, in insert_query
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 910, in execute_sql
    cursor.execute(sql, params)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/backends/util.py", line 40, in execute
    return self.cursor.execute(sql, params)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/backends/postgresql_psycopg2/base.py", line 52, in execute
    return self.cursor.execute(query, args)
IntegrityError: duplicate key value violates unique constraint "lrs_activity_activity_id_key"
DETAIL:  Key (activity_id)=(http://blahblah/someactivity) already exists.
```

On the other hand, if we make the same calls one after the other with a delay between them, we do not get these errors.

**Comment by ljwolford** [comment]
I'm not sure how you guys were setting up your async tests, but I tried to duplicate the issue and wasn't able to. I added these try excepts in ActivityManager.py to try and fix the solution:

```
    def populate(self, the_object):
        activity_id = the_object['id']
        can_define = False
        # Try to get activity
        try:
            act = Activity.objects.get(activity_id=activity_id)
        except Activity.DoesNotExist:
            if self.define_permission:
                can_define = True
                # If activity DNE and can define - create activity with auth
                try:
                    self.Activity, act_created = Activity.objects.get_or_create(activity_id=activity_id, authority=self.auth)       
                except IntegrityError:
                    self.Activity = Activity.objects.get(activity_id=activity_id, authority=self.auth)
                    act_created = False
            else:
                # If activity DNE and cannot define - create activity without auth
                try:
                    self.Activity, act_created = Activity.objects.get_or_create(activity_id=activity_id)
                except IntegrityError:
                    self.Activity = Activity.objects.get(activity_id=activity_id)
                    act_created = False
```

Please try that out and let me know if it works or not. You can either copy it from here or I can put it in a test branch. Thanks.

**Comment by BugVito** [comment]
Hey Lou,
We did something similar-ish. We cheated a little by recalling the method, and putting the try/except around the entire if/else.  I am concerned about having to use transaction.rollback() though...

I will try to get time to create a proper test for your purposes.

Here's the patch that I am using.

```
diff --git a/lrs/managers/ActivityManager.py b/lrs/managers/ActivityManager.py
index 2bf6505..7edaa5e 100644
--- a/lrs/managers/ActivityManager.py
+++ b/lrs/managers/ActivityManager.py
@@ -1,4 +1,6 @@
 from ..models import Activity
+from django.db import transaction
+from django.db import IntegrityError
 
 class ActivityManager():
     def __init__(self, data, auth=None, define=True):
@@ -50,21 +52,29 @@ class ActivityManager():
                 self.Activity.activity_definition_description = act_def['description']
         self.Activity.save()
 
-    def populate(self, the_object):
+    def populate(self, the_object, retry=True):
         activity_id = the_object['id']
         can_define = False
         # Try to get activity
         try:
             act = Activity.objects.get(activity_id=activity_id)
         except Activity.DoesNotExist:
-            act_created = True
-            if self.define_permission:
-                # If activity DNE and can define - create activity with auth
-                self.Activity = Activity.objects.create(activity_id=activity_id, authority=self.auth)
-                can_define = True
-            else:
-                # If activity DNE and cannot define - create activity without auth
-                self.Activity = Activity.objects.create(activity_id=activity_id)
+            try:
+                act_created = True
+                if self.define_permission:
+                    # If activity DNE and can define - create activity with auth
+                    self.Activity = Activity.objects.create(activity_id=activity_id, authority=self.auth)
+                    can_define = True
+                else:
+                    # If activity DNE and cannot define - create activity without auth
+                    self.Activity = Activity.objects.create(activity_id=activity_id)
+            except IntegrityError as e:
+                #Activity may have been created by a queued request, by the time this happened, retry
+                if retry:
+                    transaction.rollback()
+                    self.populate(the_object, retry=False)
+                else:
+                    raise e
         # activity already exists
         else:
             self.Activity = act
@@ -91,4 +101,4 @@ class ActivityManager():
             self.update_activity_definition(activity_definition)
         # If there is an incoming definition for an activity and the activity didn't exist yet and the user can update create the definition
         elif activity_definition and can_define and act_created:
-            self.create_activity_definition(activity_definition)
\ No newline at end of file
+            self.create_activity_definition(activity_definition)
diff --git a/lrs/models.py b/lrs/models.py
index 364d4fe..268f677 100644
--- a/lrs/models.py
+++ b/lrs/models.py
@@ -7,6 +7,8 @@ from django.db import models
 from django.contrib.auth.models import User
 from django.core.files.storage import FileSystemStorage
 from django.utils.timezone import utc
+from django.db import transaction
+from django.db import IntegrityError
 
 from oauth_provider.consts import MAX_URL_LENGTH
 
@@ -47,7 +49,7 @@ class Verb(models.Model):
         return json.dumps(self.to_dict())
 
 class AgentManager(models.Manager):
-    def retrieve_or_create(self, **kwargs):
+    def retrieve_or_create(self, retry=True, **kwargs):
         agent_ifps_can_only_be_one = ['mbox', 'mbox_sha1sum', 'account', 'openid']
         ifp_sent = [a for a in agent_ifps_can_only_be_one if kwargs.get(a, None) != None]        
         is_group = kwargs.get('objectType', None) == "Group"
@@ -78,9 +80,16 @@ class AgentManager(models.Manager):
                 agent = Agent.objects.filter(**ifp_dict)[0]
                 created = False
             except IndexError:
-                # If DNE create the agent based off of kwargs (kwargs now includes account_homePage and account_name fields)
-                agent = Agent.objects.create(**kwargs)
-                created = True
+                try:
+                    # If DNE create the agent based off of kwargs (kwargs now includes account_homePage and account_name fields)
+                    agent = Agent.objects.create(**kwargs)
+                    created = True
+                except IntegrityError as e:
+                    if retry:
+                        transaction.rollback()
+                        return self.retrieve_or_create(retry=False,**kwargs)
+                    else:
+                        raise e
 
             # For identified groups with members
             if is_group and has_member:
```

**Comment by ljwolford** [comment]
hmmm i know that transaction will be nested under the transaction decorator at the view level but i'm not sure what would be all included with this rollback. thanks again for submitting an issue with this and working with it, it's greatly appreciated! keep me updated

**Comment by BugVito** [comment]
I'm running out of steam for trying to have a proper TestCase, or even trying to use scripts to duplicate this issue.
I'm not quitting yet, but here's our scenario:
Apache using mod_wsgi, running a pyramid project.  The GUI ends up making ajax calls to the "pyramid" backend, and each of those requests may call the LRS to store the data.

I'll try something different to recreate the issue, hopefully in a testable format, but I can confirm that the patch above fixed our issues, and so far we haven't lost any statements.

**Comment by ljwolford** [comment]
Alright...I can keep trying on my end as well with both patches.

**Comment by ljwolford** [comment]
So I wrote a quick, naive, async test in JS just to see if I could get the LRS to throw the IntegrityErrors. Once it was consistently throwing those, I added the IntegrityErrors checks in the latest pull request on the webhooks branch. I will push that out to our hosted LRS and will monitor for awhile to make sure everything is still OK. Once it's OK I'll push it all to the master branch.

The only reason I went with this patch was because I was trying to keep it as simple as possible and not use any type of explicit transaction calls. If this doesn't work however, I will try your patch instead.

Once again, thanks for the work!

**Comment by BugVito** [comment]
You are the man! And I'm also not a big fan of aborting transactions without being 100% confident that it cannot affect anything else.
I did beat myself up a few hours for trying to duplicate the problem.  The closest I got, I froze my docker image...

So far, no issues though, on 4 servers.

I'm very sorry that I didn't have much time to dedicate to creating a proper test.   As for your fix, You've got my +1.
